--- a/net/minecraft/entity/projectile/AbstractArrowEntity.java
+++ b/net/minecraft/entity/projectile/AbstractArrowEntity.java
@@ -130,7 +_,7 @@
 
       BlockPos blockpos = this.func_233580_cy_();
       BlockState blockstate = this.field_70170_p.func_180495_p(blockpos);
-      if (!blockstate.func_196958_f() && !flag) {
+      if (!blockstate.isAir(this.field_70170_p, blockpos) && !flag) {
          VoxelShape voxelshape = blockstate.func_196952_d(this.field_70170_p, blockpos);
          if (!voxelshape.func_197766_b()) {
             Vector3d vector3d1 = this.func_213303_ch();
@@ -184,7 +_,7 @@
                }
             }
 
-            if (raytraceresult != null && !flag) {
+            if (raytraceresult != null && raytraceresult.func_216346_c() != RayTraceResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
                this.func_70227_a(raytraceresult);
                this.field_70160_al = true;
             }
@@ -262,7 +_,7 @@
 
    protected void func_225516_i_() {
       ++this.field_70252_j;
-      if (this.field_70252_j >= 1200) {
+      if (this.field_70252_j >= (field_70251_a == PickupStatus.CREATIVE_ONLY ? field_70170_p.paperConfig.creativeArrowDespawnRate : (field_70251_a == PickupStatus.DISALLOWED ? field_70170_p.paperConfig.nonPlayerArrowDespawnRate : ((this instanceof TridentEntity) ? field_70170_p.spigotConfig.tridentDespawnRate : field_70170_p.spigotConfig.arrowDespawnRate)))) { // Spigot // Paper - TODO: Extract this to init?
          this.func_70106_y();
       }
 
