--- a/net/minecraft/world/server/TicketType.java
+++ b/net/minecraft/world/server/TicketType.java
@@ -9,7 +_,7 @@
 public class TicketType<T> {
    private final String field_219495_h;
    private final Comparator<T> field_219496_i;
-   private final long field_223186_k;
+   public final long field_223186_k;
    public static final TicketType<Unit> field_219488_a = func_219484_a("start", (p_219486_0_, p_219486_1_) -> {
       return 0;
    });
@@ -22,6 +_,14 @@
    public static final TicketType<BlockPos> field_219493_f = func_223183_a("portal", Vector3i::compareTo, 300);
    public static final TicketType<Integer> field_223185_g = func_223183_a("post_teleport", Integer::compareTo, 5);
    public static final TicketType<ChunkPos> field_219494_g = func_223183_a("unknown", Comparator.comparingLong(ChunkPos::func_201841_a), 1);
+
+   public static final TicketType<Unit> PLUGIN = func_219484_a("plugin", (a, b) -> 0); // CraftBukkit
+   public static final TicketType<Long> LOGIN = func_223183_a("login", Long::compareTo, 100); // Paper
+   public static final TicketType<Long> FUTURE_AWAIT = func_219484_a("future_await", Long::compareTo); // Paper
+   public static final TicketType<Long> ASYNC_LOAD = func_219484_a("async_load", Long::compareTo); // Paper
+   public static final TicketType<ChunkPos> PRIORITY = func_223183_a("priority", Comparator.comparingLong(ChunkPos::func_201841_a), 300); // Paper
+   public static final TicketType<ChunkPos> URGENT = func_223183_a("urgent", Comparator.comparingLong(ChunkPos::func_201841_a), 300); // Paper
+   public static final TicketType<Long> DELAY_UNLOAD = func_223183_a("delay_unload", Long::compareTo, 300); // Paper
 
    public static <T> TicketType<T> func_219484_a(String p_219484_0_, Comparator<T> p_219484_1_) {
       return new TicketType<>(p_219484_0_, p_219484_1_, 0L);
