--- a/net/minecraft/util/concurrent/ThreadTaskExecutor.java
+++ b/net/minecraft/util/concurrent/ThreadTaskExecutor.java
@@ -74,6 +_,14 @@
 
    }
 
+   // Paper start
+   public void scheduleOnMain(Runnable r0) {
+      // postToMainThread does not work the same as older versions of mc
+      // This method is actually used to create a TickTask, which can then be posted onto main
+      this.func_212871_a_(this.func_212875_d_(r0));
+   }
+   // Paper end
+
    public void func_212871_a_(R p_212871_1_) {
       this.field_213173_d.add(p_212871_1_);
       LockSupport.unpark(this.func_213170_ax());
@@ -93,7 +_,7 @@
       this.field_213173_d.clear();
    }
 
-   protected void func_213160_bf() {
+   public void func_213160_bf() { // Paper - protected -> public
       while(this.func_213168_p()) {
       }
 
@@ -135,6 +_,7 @@
       try {
          p_213166_1_.run();
       } catch (Exception exception) {
+         if (exception.getCause() instanceof ThreadDeath) throw exception; // Paper
          field_213172_c.fatal("Error executing task on {}", this.func_213142_bd(), exception);
       }
 
