--- a/net/minecraft/util/math/vector/Vector3i.java
+++ b/net/minecraft/util/math/vector/Vector3i.java
@@ -25,6 +_,15 @@
    private int field_177960_b;
    private int field_177961_c;
 
+   // Paper start
+   public boolean isValidLocation() {
+      return func_177958_n() >= -30000000 && func_177952_p() >= -30000000 && func_177958_n() < 30000000 && func_177952_p() < 30000000 && func_177956_o() >= 0 && func_177956_o() < 256;
+   }
+   public boolean isInvalidYLocation() {
+      return field_177960_b < 0 || field_177960_b >= 256;
+   }
+   // Paper end
+
    public Vector3i(int p_i46007_1_, int p_i46007_2_, int p_i46007_3_) {
       this.field_177962_a = p_i46007_1_;
       this.field_177960_b = p_i46007_2_;
