--- a/net/minecraft/util/FoodStats.java
+++ b/net/minecraft/util/FoodStats.java
@@ -59,7 +_,7 @@
          ++this.field_75123_d;
          if (this.field_75123_d >= 80) {
             p_75118_1_.func_70691_i(1.0F);
-            this.func_75113_a(6.0F);
+            this.func_75113_a(p_75118_1_.field_70170_p.spigotConfig.regenExhaustion);
             this.field_75123_d = 0;
          }
       } else if (this.field_75127_a <= 0) {
