--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -174,7 +_,7 @@
       throw p_229756_0_ instanceof RuntimeException ? (RuntimeException)p_229756_0_ : new RuntimeException(p_229756_0_);
    }
 
-   private static void func_240983_a_(Thread p_240983_0_, Throwable p_240983_1_) {
+   public static void func_240983_a_(Thread p_240983_0_, Throwable p_240983_1_) {
       func_229757_c_(p_240983_1_);
       if (p_240983_1_ instanceof CompletionException) {
          p_240983_1_ = p_240983_1_.getCause();
@@ -200,7 +_,7 @@
       try {
          type = DataFixesManager.func_210901_a().getSchema(DataFixUtils.makeKey(SharedConstants.func_215069_a().getWorldVersion())).getChoiceType(p_240990_0_, p_240990_1_);
       } catch (IllegalArgumentException illegalargumentexception) {
-         field_195650_a.error("No data fixer registered for {}", (Object)p_240990_1_);
+         field_195650_a.debug("No data fixer registered for {}", (Object)p_240990_1_);
          if (SharedConstants.field_206244_b) {
             throw illegalargumentexception;
          }
