--- a/net/minecraft/util/WorldOptimizer.java
+++ b/net/minecraft/util/WorldOptimizer.java
@@ -117,9 +_,9 @@
                      CompoundNBT compoundnbt = chunkloader.func_227078_e_(chunkpos);
                      if (compoundnbt != null) {
                         int j = ChunkLoader.func_219165_a(compoundnbt);
-                        CompoundNBT compoundnbt1 = chunkloader.func_235968_a_(registrykey2, () -> {
+                        CompoundNBT compoundnbt1 = chunkloader.upgradeChunkTag(registrykey2, () -> {
                            return this.field_219960_q;
-                        }, compoundnbt);
+                        }, compoundnbt, chunkpos, null);
                         CompoundNBT compoundnbt2 = compoundnbt1.func_74775_l("Level");
                         ChunkPos chunkpos1 = new ChunkPos(compoundnbt2.func_74762_e("xPos"), compoundnbt2.func_74762_e("zPos"));
                         if (!chunkpos1.equals(chunkpos)) {
