--- a/net/minecraft/network/PacketThreadUtil.java
+++ b/net/minecraft/network/PacketThreadUtil.java
@@ -1,6 +_,10 @@
 package net.minecraft.network;
 
+import net.minecraft.network.play.ServerPlayNetHandler;
+import net.minecraft.network.play.server.SDisconnectPacket;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.concurrent.ThreadTaskExecutor;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.world.server.ServerWorld;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -15,8 +_,25 @@
    public static <T extends INetHandler> void func_218797_a(IPacket<T> p_218797_0_, T p_218797_1_, ThreadTaskExecutor<?> p_218797_2_) throws ThreadQuickExitException {
       if (!p_218797_2_.func_213162_bc()) {
          p_218797_2_.execute(() -> {
+            if (MinecraftServer.getServer().func_71241_aa() || (p_218797_1_ instanceof ServerPlayNetHandler && ((ServerPlayNetHandler) p_218797_1_).processedDisconnect)) return; // CraftBukkit, MC-142590
             if (p_218797_1_.func_147298_b().func_150724_d()) {
-               p_218797_0_.func_148833_a(p_218797_1_);
+               try {
+                  p_218797_0_.func_148833_a(p_218797_1_);
+               } catch (Exception e) {
+                  NetworkManager networkmanager = p_218797_1_.func_147298_b();
+                  if (networkmanager.getPlayer() != null) {
+                     field_225384_a.error("Error whilst processing packet {} for {}[{}]", p_218797_0_, networkmanager.getPlayer().func_200200_C_(), networkmanager.func_74430_c(), e);
+                  } else {
+                     field_225384_a.error("Error whilst processing packet {} for connection from {}", p_218797_0_, networkmanager.func_74430_c(), e);
+                  }
+                  StringTextComponent error = new StringTextComponent("Packet processing error");
+                  networkmanager.func_201058_a(new SDisconnectPacket(error), (future) -> {
+                     networkmanager.func_150718_a(error);
+                  });
+                  networkmanager.func_150721_g();
+               }
+               // Paper end
+
             } else {
                field_225384_a.debug("Ignoring packet due to disconnection: " + p_218797_0_);
             }
