--- a/net/minecraft/world/chunk/storage/ChunkLoader.java
+++ b/net/minecraft/world/chunk/storage/ChunkLoader.java
@@ -11,32 +_,87 @@
 import net.minecraft.util.SharedConstants;
 import net.minecraft.util.datafix.DefaultTypeReferences;
 import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import net.minecraft.world.chunk.ChunkStatus;
 import net.minecraft.world.gen.feature.structure.LegacyStructureDataUtil;
+import net.minecraft.world.server.ChunkManager;
+import net.minecraft.world.server.ServerChunkProvider;
+import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.storage.DimensionSavedDataManager;
 
 public class ChunkLoader implements AutoCloseable {
-   private final IOWorker field_227077_a_;
+   //private final IOWorker worker; nuke IOWorker
    protected final DataFixer field_219168_b;
    @Nullable
-   private LegacyStructureDataUtil field_219167_a;
+   private volatile LegacyStructureDataUtil field_219167_a; // Paper - async chunk loading
+
+   private final Object persistentDataLock = new Object(); // Paper
+   public final RegionFileCache regionFileCache;
 
    public ChunkLoader(File p_i231889_1_, DataFixer p_i231889_2_, boolean p_i231889_3_) {
+      this.regionFileCache = new RegionFileCache(p_i231889_1_, p_i231889_3_); // Paper - nuke IOWorker
       this.field_219168_b = p_i231889_2_;
-      this.field_227077_a_ = new IOWorker(p_i231889_1_, p_i231889_3_, "chunk");
-   }
-
-   public CompoundNBT func_235968_a_(RegistryKey<World> p_235968_1_, Supplier<DimensionSavedDataManager> p_235968_2_, CompoundNBT p_235968_3_) {
+      //this.worker = new IOWorker(p_i231889_1_, p_i231889_3_, "chunk");
+   }
+
+   private boolean check(ServerChunkProvider cps, int x, int z) throws IOException {
+      ChunkPos pos = new ChunkPos(x, z);
+      if (cps != null) {
+         //com.google.common.base.Preconditions.checkState(org.bukkit.Bukkit.isPrimaryThread(), "primary thread"); // Paper - this function is now MT-Safe
+         if (cps.getChunkAtIfCachedImmediately(x, z) != null) { // Paper - isLoaded is a ticket level check, not a chunk loaded check!
+            return true;
+         }
+      }
+
+
+      // Paper start - prioritize
+      CompoundNBT nbt = cps == null ? func_227078_e_(pos) :
+              com.destroystokyo.paper.io.PaperFileIOThread.Holder.INSTANCE.loadChunkData((ServerWorld)cps.field_73251_h, x, z,
+                      com.destroystokyo.paper.io.PrioritizedTaskQueue.HIGHER_PRIORITY, false, true).chunkData;
+      // Paper end
+
+      if (nbt != null) {
+         CompoundNBT level = nbt.func_74775_l("Level");
+         if (level.func_74767_n("TerrainPopulated")) {
+            return true;
+         }
+
+         ChunkStatus status = ChunkStatus.func_222591_a(level.func_74779_i("Status"));
+         if (status != null && status.func_209003_a(ChunkStatus.field_222613_i)) {
+            return true;
+         }
+      }
+
+      return false;
+   }
+
+   public CompoundNBT upgradeChunkTag(RegistryKey<World> p_235968_1_, Supplier<DimensionSavedDataManager> p_235968_2_, CompoundNBT p_235968_3_, ChunkPos pos, @Nullable IWorld generatoraccess) throws IOException {
       int i = func_219165_a(p_235968_3_);
       int j = 1493;
+
+//      // CraftBukkit start
+//      if (i < 1466) {
+//         CompoundNBT level = p_235968_3_.getCompound("Level");
+//         if (level.getBoolean("TerrainPopulated") && !level.getBoolean("LightPopulated")) {
+//            ServerChunkProvider cps = (generatoraccess == null) ? null : ((ServerWorld) generatoraccess).getChunkSource();
+//            if (check(cps, pos.x - 1, pos.z) && check(cps, pos.x - 1, pos.z - 1) && check(cps, pos.x, pos.z - 1)) {
+//               level.putBoolean("LightPopulated", true);
+//            }
+//         }
+//      }
+//      // CraftBukkit end
+
       if (i < 1493) {
          p_235968_3_ = NBTUtil.func_210821_a(this.field_219168_b, DefaultTypeReferences.CHUNK, p_235968_3_, i, 1493);
          if (p_235968_3_.func_74775_l("Level").func_74767_n("hasLegacyStructureData")) {
-            if (this.field_219167_a == null) {
-               this.field_219167_a = LegacyStructureDataUtil.func_236992_a_(p_235968_1_, p_235968_2_.get());
-            }
+            synchronized (this.persistentDataLock) { // Paper - Async chunk loading
+               if (this.field_219167_a == null) {
+                  this.field_219167_a = LegacyStructureDataUtil.func_236992_a_(p_235968_1_, p_235968_2_.get());
+               }
 
-            p_235968_3_ = this.field_219167_a.func_212181_a(p_235968_3_);
+               p_235968_3_ = this.field_219167_a.func_212181_a(p_235968_3_);
+            } // Paper - Async chunk loading
          }
       }
 
@@ -54,22 +_,31 @@
 
    @Nullable
    public CompoundNBT func_227078_e_(ChunkPos p_227078_1_) throws IOException {
-      return this.field_227077_a_.func_227090_a_(p_227078_1_);
+      return this.regionFileCache.func_219099_e(p_227078_1_);
    }
 
-   public void func_219100_a(ChunkPos p_219100_1_, CompoundNBT p_219100_2_) {
-      this.field_227077_a_.func_227093_a_(p_219100_1_, p_219100_2_);
+   public void func_219100_a(ChunkPos p_219100_1_, CompoundNBT p_219100_2_) throws IOException {
+      // Paper start
+      if (!p_219100_1_.equals(ChunkSerializer.getChunkCoordinate(p_219100_2_))) {
+         String world = (this instanceof ChunkManager) ? ((ChunkManager)this).field_219255_i.func_234923_W_().func_240901_a_().toString() : null;
+         throw new IllegalArgumentException("Chunk coordinate and serialized data do not have matching coordinates, trying to serialize coordinate " + p_219100_1_.toString()
+                 + " but compound says coordinate is " + ChunkSerializer.getChunkCoordinate(p_219100_2_).toString() + (world == null ? " for an unknown world" : (" for world: " + world)));
+      }
+      // Paper end
+      this.regionFileCache.func_219100_a(p_219100_1_, p_219100_2_);
       if (this.field_219167_a != null) {
-         this.field_219167_a.func_208216_a(p_219100_1_.func_201841_a());
+         synchronized (this.persistentDataLock) {
+            this.field_219167_a.func_208216_a(p_219100_1_.func_201841_a());
+         }
       }
 
    }
 
-   public void func_227079_i_() {
-      this.field_227077_a_.func_227088_a_().join();
-   }
+//   public void flushWorker() {
+//      this.worker.synchronize().join();
+//   }
 
    public void close() throws IOException {
-      this.field_227077_a_.close();
+      this.regionFileCache.close();
    }
 }
