--- a/net/minecraft/entity/item/ItemEntity.java
+++ b/net/minecraft/entity/item/ItemEntity.java
@@ -1,5 +_,7 @@
 package net.minecraft.entity.item;
 
+import java.util.Iterator;
+import java.util.List;
 import java.util.Objects;
 import java.util.UUID;
 import javax.annotation.Nullable;
@@ -16,6 +_,7 @@
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.network.play.server.SSpawnObjectPacket;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.DamageSource;
@@ -38,6 +_,34 @@
    private UUID field_145801_f;
    private UUID field_145802_g;
    public final float field_70290_d;
+   private int lastTick = MinecraftServer.currentTick - 1; // CraftBukkit
+   /**
+    * The maximum age of this EntityItem.  The item is expired once this is reached.
+    */
+   public int lifespan = 6000;
+
+   // Spigot start - copied from above
+   @Override
+   public void inactiveTick() {
+      // CraftBukkit start - Use wall time for pickup and despawn timers
+      int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+      if (this.field_145804_b != 32767) this.field_145804_b -= elapsedTicks;
+      this.field_145804_b = Math.max(0, this.field_145804_b); // Paper - don't go below 0
+      if (this.field_70292_b != -32768) this.field_70292_b += elapsedTicks;
+      this.lastTick = MinecraftServer.currentTick;
+      // CraftBukkit end
+
+      if (!this.field_70170_p.field_72995_K && this.field_70292_b >= this.getDespawnRate()) { // Spigot // Paper
+         this.func_70106_y();
+      }
+   }
+   // Spigot end
+
+   // Paper start
+   public int getDespawnRate(){
+      return field_70170_p.spigotConfig.itemDespawnRate;
+   }
+   // Paper end
 
    public ItemEntity(EntityType<? extends ItemEntity> p_i50217_1_, World p_i50217_2_) {
       super(p_i50217_1_, p_i50217_2_);
@@ -54,6 +_,7 @@
    public ItemEntity(World p_i1710_1_, double p_i1710_2_, double p_i1710_4_, double p_i1710_6_, ItemStack p_i1710_8_) {
       this(p_i1710_1_, p_i1710_2_, p_i1710_4_, p_i1710_6_);
       this.func_92058_a(p_i1710_8_);
+      this.lifespan = (p_i1710_8_.func_77973_b() == null ? 6000 : p_i1710_8_.getEntityLifespan(p_i1710_1_));
    }
 
    @OnlyIn(Dist.CLIENT)
@@ -74,13 +_,21 @@
    }
 
    public void func_70071_h_() {
+      if (func_92059_d().onEntityItemUpdate(this)) return;
       if (this.func_92059_d().func_190926_b()) {
          this.func_70106_y();
       } else {
          super.func_70071_h_();
-         if (this.field_145804_b > 0 && this.field_145804_b != 32767) {
-            --this.field_145804_b;
-         }
+         // CraftBukkit start - Use wall time for pickup and despawn timers
+         int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+         if (this.field_145804_b != 32767) this.field_145804_b -= elapsedTicks;
+         this.field_145804_b = Math.max(0, this.field_145804_b); // Paper - don't go below 0
+         if (this.field_70292_b != -32768) this.field_70292_b += elapsedTicks;
+         this.lastTick = MinecraftServer.currentTick;
+         // CraftBukkit end
+//         if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
+//            --this.pickupDelay;
+//         }
 
          this.field_70169_q = this.func_226277_ct_();
          this.field_70167_r = this.func_226278_cu_();
@@ -108,7 +_,7 @@
             this.func_213315_a(MoverType.SELF, this.func_213322_ci());
             float f1 = 0.98F;
             if (this.field_70122_E) {
-               f1 = this.field_70170_p.func_180495_p(new BlockPos(this.func_226277_ct_(), this.func_226278_cu_() - 1.0D, this.func_226281_cx_())).func_177230_c().func_208618_m() * 0.98F;
+               f1 = this.field_70170_p.func_180495_p(new BlockPos(this.func_226277_ct_(), this.func_226278_cu_() - 1.0D, this.func_226281_cx_())).getSlipperiness(field_70170_p, new BlockPos(this.func_226277_ct_(), this.func_226278_cu_() - 1.0D, this.func_226281_cx_()), this) * 0.98F;
             }
 
             this.func_213317_d(this.func_213322_ci().func_216372_d((double)f1, 0.98D, (double)f1));
@@ -144,7 +_,14 @@
             }
          }
 
-         if (!this.field_70170_p.field_72995_K && this.field_70292_b >= 6000) {
+         ItemStack item = this.func_92059_d();
+         if (!this.field_70170_p.field_72995_K && this.field_70292_b >= lifespan) {
+             int hook = net.minecraftforge.event.ForgeEventFactory.onItemExpire(this, item);
+             if (hook < 0) this.func_70106_y();
+             else          this.lifespan += hook;
+         }
+
+         if (item.func_190926_b()) {
             this.func_70106_y();
          }
 
@@ -163,17 +_,26 @@
 
    private void func_85054_d() {
       if (this.func_213857_z()) {
-         for(ItemEntity itementity : this.field_70170_p.func_175647_a(ItemEntity.class, this.func_174813_aQ().func_72314_b(0.5D, 0.0D, 0.5D), (p_213859_1_) -> {
+
+         // Spigot start
+         double radius = field_70170_p.spigotConfig.itemMerge;
+         List<ItemEntity> list = this.field_70170_p.func_175647_a(ItemEntity.class, this.func_174813_aQ().func_72314_b(radius, radius, radius), (p_213859_1_) -> {
+            // Spigot end
             return p_213859_1_ != this && p_213859_1_.func_213857_z();
-         })) {
-            if (itementity.func_213857_z()) {
-               this.func_226530_a_(itementity);
+         });
+
+         Iterator iterator = list.iterator();
+
+         while (iterator.hasNext()) {
+            ItemEntity entityitem = (ItemEntity) iterator.next();
+
+            if (entityitem.func_213857_z()) {
+               this.func_226530_a_(entityitem);
                if (this.field_70128_L) {
                   break;
                }
             }
          }
-
       }
    }
 
@@ -202,6 +_,8 @@
          return false;
       } else if (p_226532_1_.func_77942_o() ^ p_226532_0_.func_77942_o()) {
          return false;
+      } else if (!p_226532_0_.areCapsCompatible(p_226532_1_)) {
+         return false;
       } else {
          return !p_226532_1_.func_77942_o() || p_226532_1_.func_77978_p().equals(p_226532_0_.func_77978_p());
       }
@@ -235,6 +_,7 @@
    }
 
    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_) {
+      if (this.field_70170_p.field_72995_K || this.field_70128_L) return false; //Forge: Fixes MC-53850
       if (this.func_180431_b(p_70097_1_)) {
          return false;
       } else if (!this.func_92059_d().func_190926_b() && this.func_92059_d().func_77973_b() == Items.field_151156_bN && p_70097_1_.func_94541_c()) {
@@ -256,6 +_,7 @@
       p_213281_1_.func_74777_a("Health", (short)this.field_70291_e);
       p_213281_1_.func_74777_a("Age", (short)this.field_70292_b);
       p_213281_1_.func_74777_a("PickupDelay", (short)this.field_145804_b);
+      p_213281_1_.func_74768_a("Lifespan", lifespan);
       if (this.func_200214_m() != null) {
          p_213281_1_.func_186854_a("Thrower", this.func_200214_m());
       }
@@ -276,6 +_,7 @@
       if (p_70037_1_.func_74764_b("PickupDelay")) {
          this.field_145804_b = p_70037_1_.func_74765_d("PickupDelay");
       }
+      if (p_70037_1_.func_74764_b("Lifespan")) lifespan = p_70037_1_.func_74762_e("Lifespan");
 
       if (p_70037_1_.func_186855_b("Owner")) {
          this.field_145802_g = p_70037_1_.func_186857_a("Owner");
@@ -295,10 +_,18 @@
 
    public void func_70100_b_(PlayerEntity p_70100_1_) {
       if (!this.field_70170_p.field_72995_K) {
+         if (this.field_145804_b > 0) return;
          ItemStack itemstack = this.func_92059_d();
          Item item = itemstack.func_77973_b();
          int i = itemstack.func_190916_E();
-         if (this.field_145804_b == 0 && (this.field_145802_g == null || this.field_145802_g.equals(p_70100_1_.func_110124_au())) && p_70100_1_.field_71071_by.func_70441_a(itemstack)) {
+
+         int hook = net.minecraftforge.event.ForgeEventFactory.onItemPickup(this, p_70100_1_);
+         if (hook < 0) return;
+
+         ItemStack copy = itemstack.func_77946_l();
+         if (this.field_145804_b == 0 && (this.field_145802_g == null || lifespan - this.field_70292_b <= 200 || this.field_145802_g.equals(p_70100_1_.func_110124_au())) && (hook == 1 || i <= 0 || p_70100_1_.field_71071_by.func_70441_a(itemstack))) {
+            copy.func_190920_e(copy.func_190916_E() - func_92059_d().func_190916_E());
+            net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerItemPickupEvent(p_70100_1_, this, copy);
             p_70100_1_.func_71001_a(this, i);
             if (itemstack.func_190926_b()) {
                this.func_70106_y();
@@ -322,8 +_,8 @@
    }
 
    @Nullable
-   public Entity func_241206_a_(ServerWorld p_241206_1_) {
-      Entity entity = super.func_241206_a_(p_241206_1_);
+   public Entity changeDimension(ServerWorld p_241206_1_, net.minecraftforge.common.util.ITeleporter teleporter) {
+      Entity entity = super.changeDimension(p_241206_1_, teleporter);
       if (!this.field_70170_p.field_72995_K && entity instanceof ItemEntity) {
          ((ItemEntity)entity).func_85054_d();
       }
@@ -396,7 +_,7 @@
 
    public void func_174870_v() {
       this.func_174871_r();
-      this.field_70292_b = 5999;
+      this.field_70292_b = func_92059_d().getEntityLifespan(field_70170_p) - 1;
    }
 
    @OnlyIn(Dist.CLIENT)
