--- a/net/minecraft/entity/item/TNTEntity.java
+++ b/net/minecraft/entity/item/TNTEntity.java
@@ -53,6 +_,7 @@
    }
 
    public void func_70071_h_() {
+      if (field_70170_p.spigotConfig.currentPrimedTnt++ > field_70170_p.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
       if (!this.func_189652_ae()) {
          this.func_213317_d(this.func_213322_ci().func_72441_c(0.0D, -0.04D, 0.0D));
       }
