--- a/net/minecraft/scoreboard/ScoreboardSaveData.java
+++ b/net/minecraft/scoreboard/ScoreboardSaveData.java
@@ -161,6 +_,7 @@
       ListNBT listnbt = new ListNBT();
 
       for(ScorePlayerTeam scoreplayerteam : this.field_96507_a.func_96525_g()) {
+         if (!com.destroystokyo.paper.PaperConfig.saveEmptyScoreboardTeams && scoreplayerteam.func_96670_d().isEmpty()) continue; // Paper
          CompoundNBT compoundnbt = new CompoundNBT();
          compoundnbt.func_74778_a("Name", scoreplayerteam.func_96661_b());
          compoundnbt.func_74778_a("DisplayName", ITextComponent.Serializer.func_150696_a(scoreplayerteam.func_96669_c()));
