--- a/net/minecraft/world/gen/feature/structure/EndCityPieces.java
+++ b/net/minecraft/world/gen/feature/structure/EndCityPieces.java
@@ -254,12 +_,12 @@
                LockableLootTileEntity.func_195479_a(p_186175_3_, p_186175_4_, blockpos, LootTables.field_186421_c);
             }
          } else if (p_186175_1_.startsWith("Sentry")) {
-            ShulkerEntity shulkerentity = EntityType.field_200738_ad.func_200721_a(p_186175_3_.func_201672_e());
+            ShulkerEntity shulkerentity = EntityType.field_200738_ad.func_200721_a(p_186175_3_.getLevel());
             shulkerentity.func_70107_b((double)p_186175_2_.func_177958_n() + 0.5D, (double)p_186175_2_.func_177956_o() + 0.5D, (double)p_186175_2_.func_177952_p() + 0.5D);
             shulkerentity.func_184694_g(p_186175_2_);
             p_186175_3_.func_217376_c(shulkerentity);
          } else if (p_186175_1_.startsWith("Elytra")) {
-            ItemFrameEntity itemframeentity = new ItemFrameEntity(p_186175_3_.func_201672_e(), p_186175_2_, this.field_186182_e.func_185831_a(Direction.SOUTH));
+            ItemFrameEntity itemframeentity = new ItemFrameEntity(p_186175_3_.getLevel(), p_186175_2_, this.field_186182_e.func_185831_a(Direction.SOUTH));
             itemframeentity.func_174864_a(new ItemStack(Items.field_185160_cR), false);
             p_186175_3_.func_217376_c(itemframeentity);
          }
