--- a/net/minecraft/entity/ai/goal/Goal.java
+++ b/net/minecraft/entity/ai/goal/Goal.java
@@ -1,9 +_,12 @@
 package net.minecraft.entity.ai.goal;
 
+import com.destroystokyo.paper.util.set.OptimizedSmallEnumSet;
+
 import java.util.EnumSet;
 
 public abstract class Goal {
    private final EnumSet<Goal.Flag> field_220687_a = EnumSet.noneOf(Goal.Flag.class);
+   private final OptimizedSmallEnumSet<Goal.Flag> goalTypes = new OptimizedSmallEnumSet<>(Goal.Flag.class); // Paper - remove streams from pathfindergoalselector
 
    public abstract boolean func_75250_a();
 
@@ -35,6 +_,12 @@
 
    public EnumSet<Goal.Flag> func_220686_i() {
       return this.field_220687_a;
+   }
+
+   // Paper start - remove streams from pathfindergoalselector
+   public com.destroystokyo.paper.util.set.OptimizedSmallEnumSet<Goal.Flag> getGoalTypes() {
+      return this.goalTypes;
+      // Paper end - remove streams from pathfindergoalselector
    }
 
    public static enum Flag {
