--- a/net/minecraft/entity/ai/goal/PrioritizedGoal.java
+++ b/net/minecraft/entity/ai/goal/PrioritizedGoal.java
@@ -55,6 +_,12 @@
       return this.field_220774_a.func_220686_i();
    }
 
+   // Paper start - remove streams from pathfindergoalselector
+   public com.destroystokyo.paper.util.set.OptimizedSmallEnumSet<Goal.Flag> getGoalTypes() {
+      return this.field_220774_a.getGoalTypes();
+      // Paper end - remove streams from pathfindergoalselector
+   }
+
    public boolean func_220773_g() {
       return this.field_220776_c;
    }
