--- a/net/minecraft/entity/ai/goal/SwimGoal.java
+++ b/net/minecraft/entity/ai/goal/SwimGoal.java
@@ -9,6 +_,7 @@
 
    public SwimGoal(MobEntity p_i1624_1_) {
       this.field_75373_a = p_i1624_1_;
+      if (p_i1624_1_.field_70170_p.paperConfig.nerfedMobsShouldJump) p_i1624_1_.goalFloat = this; // Paper
       this.func_220684_a(EnumSet.of(Goal.Flag.JUMP));
       p_i1624_1_.func_70661_as().func_212239_d(true);
    }
