--- a/net/minecraft/util/palette/PalettedContainer.java
+++ b/net/minecraft/util/palette/PalettedContainer.java
@@ -66,8 +_,11 @@
    }
 
    private void func_186012_b(int p_186012_1_) {
-      if (p_186012_1_ != this.field_186024_e) {
-         this.field_186024_e = p_186012_1_;
+      setBits(p_186012_1_, false);
+   }
+   private void setBits(int bitsIn, boolean forceBits) {
+      if (bitsIn != this.field_186024_e) {
+         this.field_186024_e = bitsIn;
          if (this.field_186024_e <= 4) {
             this.field_186024_e = 4;
             this.field_186022_c = new ArrayPalette<>(this.field_205523_d, this.field_186024_e, this, this.field_205524_e);
@@ -76,6 +_,8 @@
          } else {
             this.field_186022_c = this.field_205521_b;
             this.field_186024_e = MathHelper.func_151241_e(this.field_205523_d.func_186804_a());
+            if (forceBits)
+               this.field_186024_e = bitsIn;
          }
 
          this.field_186022_c.func_186041_a(this.field_205526_g);
@@ -138,12 +_,16 @@
       this.func_210459_b();
       int i = p_186010_1_.readByte();
       if (this.field_186024_e != i) {
-         this.func_186012_b(i);
+         this.setBits(i, true); //Forge, Force bit density to fix network issues, resize below if needed.
       }
 
       this.field_186022_c.func_186038_a(p_186010_1_);
       p_186010_1_.func_186873_b(this.field_186021_b.func_188143_a());
       this.func_210460_c();
+
+      int regSize = MathHelper.func_151241_e(this.field_205523_d.func_186804_a());
+      if (this.field_186022_c == field_205521_b && this.field_186024_e != regSize) // Resize bits to fit registry.
+         this.onResize(regSize, field_205526_g);
    }
 
    public void func_186009_b(PacketBuffer p_186009_1_) {
@@ -154,7 +_,7 @@
       this.func_210460_c();
    }
 
-   public void func_222642_a(ListNBT p_222642_1_, long[] p_222642_2_) {
+   public void func_222642_a(ListNBT p_222642_1_, long[] p_222642_2_) { // Paper - synchronize
       this.func_210459_b();
       int i = Math.max(4, MathHelper.func_151241_e(p_222642_1_.size()));
       if (i != this.field_186024_e) {
@@ -184,7 +_,7 @@
       this.func_210460_c();
    }
 
-   public void func_196963_b(CompoundNBT p_196963_1_, String p_196963_2_, String p_196963_3_) {
+   public void func_196963_b(CompoundNBT p_196963_1_, String p_196963_2_, String p_196963_3_) { // Paper - synchronize
       this.func_210459_b();
       HashMapPalette<T> hashmappalette = new HashMapPalette<>(this.field_205523_d, this.field_186024_e, this.field_205522_c, this.field_205524_e, this.field_205525_f);
       T t = this.field_205526_g;
@@ -232,6 +_,14 @@
          p_225497_1_.accept(this.field_186022_c.func_186039_a(p_225499_2_.getIntKey()), p_225499_2_.getIntValue());
       });
    }
+
+   // Paper start
+   public void forEachLocation(PalettedContainer.ICountConsumer<T> datapaletteblock_a) {
+      this.field_186021_b.forEach((int location, int data) -> {
+         datapaletteblock_a.accept(this.field_186022_c.func_186039_a(data), location);
+      });
+   }
+   // Paper end
 
    @FunctionalInterface
    public interface ICountConsumer<T> {
