--- a/net/minecraft/world/gen/treedecorator/BeehiveTreeDecorator.java
+++ b/net/minecraft/world/gen/treedecorator/BeehiveTreeDecorator.java
@@ -51,7 +_,7 @@
                   int j = 2 + p_225576_2_.nextInt(2);
 
                   for(int k = 0; k < j; ++k) {
-                     BeeEntity beeentity = new BeeEntity(EntityType.field_226289_e_, p_225576_1_.func_201672_e());
+                     BeeEntity beeentity = new BeeEntity(EntityType.field_226289_e_, p_225576_1_.getLevel());
                      beehivetileentity.func_226962_a_(beeentity, false, p_225576_2_.nextInt(599));
                   }
                }
