--- a/net/minecraft/world/ICollisionReader.java
+++ b/net/minecraft/world/ICollisionReader.java
@@ -55,7 +_,9 @@
    }
 
    default boolean func_234865_b_(@Nullable Entity p_234865_1_, AxisAlignedBB p_234865_2_, Predicate<Entity> p_234865_3_) {
+      try { if (p_234865_1_ != null) p_234865_1_.collisionLoadChunks = true; // Paper
       return this.func_234867_d_(p_234865_1_, p_234865_2_, p_234865_3_).allMatch(VoxelShape::func_197766_b);
+      } finally { if (p_234865_1_ != null) p_234865_1_.collisionLoadChunks = false; } // Paper
    }
 
    Stream<VoxelShape> func_230318_c_(@Nullable Entity p_230318_1_, AxisAlignedBB p_230318_2_, Predicate<Entity> p_230318_3_);
