--- a/net/minecraft/world/IWorldWriter.java
+++ b/net/minecraft/world/IWorldWriter.java
@@ -3,6 +_,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockState;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.SpawnReason;
 import net.minecraft.util.math.BlockPos;
 
 public interface IWorldWriter {
@@ -27,4 +_,10 @@
    default boolean func_217376_c(Entity p_217376_1_) {
       return false;
    }
+
+   // CraftBukkit start
+   default boolean addFreshEntity(Entity entity, SpawnReason reason) {
+      return false;
+   }
+   // CraftBukkit end
 }
