--- a/net/minecraft/world/NextTickListEntry.java
+++ b/net/minecraft/world/NextTickListEntry.java
@@ -8,8 +_,8 @@
    private final T field_151352_g;
    public final BlockPos field_180282_a;
    public final long field_235017_b_;
-   public final TickPriority field_82754_f;
-   private final long field_77178_g;
+   public final TickPriority field_82754_f; public final TickPriority getPriority() { return this.field_82754_f; } // Paper - OBFHELPER
+   private final long field_77178_g; public final long getId() { return this.field_77178_g; } // Paper - OBFHELPER
 
    public NextTickListEntry(BlockPos p_i48977_1_, T p_i48977_2_) {
       this(p_i48977_1_, p_i48977_2_, 0L, TickPriority.NORMAL);
