--- a/net/minecraft/block/NetherPortalBlock.java
+++ b/net/minecraft/block/NetherPortalBlock.java
@@ -48,7 +_,7 @@
    }
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
-      if (p_225542_2_.func_230315_m_().func_236043_f_() && p_225542_2_.func_82736_K().func_223586_b(GameRules.field_223601_d) && p_225542_4_.nextInt(2000) < p_225542_2_.func_175659_aa().func_151525_a()) {
+      if (p_225542_2_.spigotConfig.enableZombiePigmenPortalSpawns && p_225542_2_.func_230315_m_().func_236043_f_() && p_225542_2_.func_82736_K().func_223586_b(GameRules.field_223601_d) && p_225542_4_.nextInt(2000) < p_225542_2_.func_175659_aa().func_151525_a()) {
          while(p_225542_2_.func_180495_p(p_225542_3_).func_203425_a(this)) {
             p_225542_3_ = p_225542_3_.func_177977_b();
          }
