--- a/net/minecraft/world/chunk/ChunkPrimerWrapper.java
+++ b/net/minecraft/world/chunk/ChunkPrimerWrapper.java
@@ -38,6 +_,11 @@
    }
 
    @Nullable
+   public BlockState getBlockState(final int x, final int y, final int z) {
+      return this.field_209220_a.getBlockState(x, y, z);
+   }
+
+   @Nullable
    public BlockState func_180495_p(BlockPos p_180495_1_) {
       return this.field_209220_a.func_180495_p(p_180495_1_);
    }
