--- a/net/minecraft/world/chunk/EmptyChunk.java
+++ b/net/minecraft/world/chunk/EmptyChunk.java
@@ -31,6 +_,10 @@
       super(p_i49950_1_, p_i49950_2_, new BiomeContainer(p_i49950_1_.func_241828_r().func_243612_b(Registry.field_239720_u_), field_201623_e));
    }
 
+   public BlockState getBlockState(int x, int y, int z) {
+      return Blocks.field_201940_ji.func_176223_P();
+   }
+
    public BlockState func_180495_p(BlockPos p_180495_1_) {
       return Blocks.field_201940_ji.func_176223_P();
    }
