--- a/net/minecraft/util/math/shapes/VoxelShapeSpliterator.java
+++ b/net/minecraft/util/math/shapes/VoxelShapeSpliterator.java
@@ -5,9 +_,12 @@
 import java.util.function.BiPredicate;
 import java.util.function.Consumer;
 import javax.annotation.Nullable;
+
+import com.destroystokyo.paper.util.MCUtil;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.CubeCoordinateIterator;
@@ -15,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.ICollisionReader;
 import net.minecraft.world.border.WorldBorder;
+import net.minecraft.world.gen.WorldGenRegion;
 
 public class VoxelShapeSpliterator extends AbstractSpliterator<VoxelShape> {
    @Nullable
@@ -60,26 +_,49 @@
    boolean func_234878_a_(Consumer<? super VoxelShape> p_234878_1_) {
       while(true) {
          if (this.field_234871_d_.func_218301_a()) {
-            int i = this.field_234871_d_.func_218304_b();
-            int j = this.field_234871_d_.func_218302_c();
-            int k = this.field_234871_d_.func_218303_d();
+            int i = this.field_234871_d_.func_218304_b(); final int x = i;
+            int j = this.field_234871_d_.func_218302_c(); final int y = j;
+            int k = this.field_234871_d_.func_218303_d(); final int z = k;
             int l = this.field_234871_d_.func_223473_e();
             if (l == 3) {
                continue;
             }
 
-            IBlockReader iblockreader = this.func_234876_a_(i, k);
-            if (iblockreader == null) {
-               continue;
-            }
-
-            this.field_234872_e_.func_181079_c(i, j, k);
-            BlockState blockstate = iblockreader.func_180495_p(this.field_234872_e_);
-            if (!this.field_241458_i_.test(blockstate, this.field_234872_e_) || l == 1 && !blockstate.func_215704_f() || l == 2 && !blockstate.func_203425_a(Blocks.field_196603_bb)) {
-               continue;
-            }
-
-            VoxelShape voxelshape = blockstate.func_215685_b(this.field_234874_g_, this.field_234872_e_, this.field_234870_c_);
+//            IBlockReader iblockreader = this.getChunk(i, k);
+//            if (iblockreader == null) {
+//               continue;
+//            }
+
+            // Paper start - ensure we don't load chunks
+            Entity entity = this.field_234868_a_;
+            BlockPos.Mutable blockposition_mutableblockposition = field_234872_e_;
+            boolean far = entity != null && MCUtil.distanceSq(entity.func_226277_ct_(), y, entity.func_226281_cx_(), x, y, z) > 14;
+            blockposition_mutableblockposition.func_181079_c(x, y, z);
+
+            boolean isRegionLimited = this.field_234874_g_ instanceof WorldGenRegion;
+            BlockState iblockdata = isRegionLimited ? Blocks.field_201940_ji.func_176223_P() : ((!far && entity instanceof ServerPlayerEntity) || (entity != null && entity.collisionLoadChunks)
+                    ? this.field_234874_g_.func_180495_p(blockposition_mutableblockposition)
+                    : this.field_234874_g_.getTypeIfLoaded(blockposition_mutableblockposition)
+            );
+
+            if (iblockdata == null) {
+               if (!(entity instanceof ServerPlayerEntity) || entity.field_70170_p.paperConfig.preventMovingIntoUnloadedChunks) {
+                  VoxelShape voxelshape3 = VoxelShapes.func_197881_a(far ? entity.func_174813_aQ() : new AxisAlignedBB(new BlockPos(x, y, z)));
+                  p_234878_1_.accept(voxelshape3);
+                  return true;
+               }
+               continue;
+            }
+            // Paper - moved up
+            // Paper end
+
+//            this.pos.set(i, j, k);
+//            BlockState blockstate = iblockreader.getBlockState(this.pos);
+            if (!this.field_241458_i_.test(iblockdata, this.field_234872_e_) || l == 1 && !iblockdata.func_215704_f() || l == 2 && !iblockdata.func_203425_a(Blocks.field_196603_bb)) {
+               continue;
+            }
+
+            VoxelShape voxelshape = iblockdata.func_215685_b(this.field_234874_g_, this.field_234872_e_, this.field_234870_c_);
             if (voxelshape == VoxelShapes.func_197868_b()) {
                if (!this.field_234869_b_.func_186668_a((double)i, (double)j, (double)k, (double)i + 1.0D, (double)j + 1.0D, (double)k + 1.0D)) {
                   continue;
