--- a/net/minecraft/command/CommandSource.java
+++ b/net/minecraft/command/CommandSource.java
@@ -33,6 +_,7 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.spigotmc.SpigotConfig;
 
 public class CommandSource implements ISuggestionProvider {
    public static final SimpleCommandExceptionType field_197039_a = new SimpleCommandExceptionType(new TranslationTextComponent("permissions.requires.player"));
@@ -207,7 +_,7 @@
          }
       }
 
-      if (this.field_197041_c != this.field_197047_i && this.field_197047_i.func_200252_aR().func_223586_b(GameRules.field_223608_k)) {
+      if (this.field_197041_c != this.field_197047_i && this.field_197047_i.func_200252_aR().func_223586_b(GameRules.field_223608_k) && !SpigotConfig.silentCommandBlocks) {
          this.field_197047_i.func_145747_a(itextcomponent, Util.field_240973_b_);
       }
 
