--- a/net/minecraft/tileentity/HopperTileEntity.java
+++ b/net/minecraft/tileentity/HopperTileEntity.java
@@ -89,9 +_,14 @@
          this.field_190578_g = this.field_145850_b.func_82737_E();
          if (!this.func_145888_j()) {
             this.func_145896_c(0);
-            this.func_200109_a(() -> {
+            // Spigot start
+            boolean result = this.func_200109_a(() -> {
                return func_145891_a(this);
             });
+            if (!result && this.field_145850_b.spigotConfig.hopperCheck > 1) {
+               this.func_145896_c(this.field_145850_b.spigotConfig.hopperCheck);
+            }
+            // Spigot end
          }
 
       }
@@ -110,7 +_,7 @@
             }
 
             if (flag) {
-               this.func_145896_c(8);
+               this.func_145896_c(field_145850_b.spigotConfig.hopperTransfer); // Spigot
                this.func_70296_d();
                return true;
             }
@@ -133,6 +_,7 @@
    }
 
    private boolean func_145883_k() {
+      if (net.minecraftforge.items.VanillaInventoryCodeHooks.insertHook(this)) return true;
       IInventory iinventory = this.func_145895_l();
       if (iinventory == null) {
          return false;
@@ -177,6 +_,8 @@
    }
 
    public static boolean func_145891_a(IHopper p_145891_0_) {
+      Boolean ret = net.minecraftforge.items.VanillaInventoryCodeHooks.extractHook(p_145891_0_);
+      if (ret != null) return ret;
       IInventory iinventory = func_145884_b(p_145891_0_);
       if (iinventory != null) {
          Direction direction = Direction.DOWN;
@@ -284,7 +_,7 @@
                      }
                   }
 
-                  hoppertileentity1.func_145896_c(8 - k);
+                  hoppertileentity1.func_145896_c(hoppertileentity1.field_145850_b.spigotConfig.hopperTransfer - k);
                }
             }
 
@@ -325,7 +_,7 @@
       Block block = blockstate.func_177230_c();
       if (block instanceof ISidedInventoryProvider) {
          iinventory = ((ISidedInventoryProvider)block).func_219966_a(blockstate, p_145893_0_, blockpos);
-      } else if (block.func_235695_q_()) {
+      } else if (blockstate.hasTileEntity()) {
          TileEntity tileentity = p_145893_0_.func_175625_s(blockpos);
          if (tileentity instanceof IInventory) {
             iinventory = (IInventory)tileentity;
@@ -403,5 +_,14 @@
 
    protected Container func_213906_a(int p_213906_1_, PlayerInventory p_213906_2_) {
       return new HopperContainer(p_213906_1_, p_213906_2_, this);
+   }
+
+   @Override
+   protected net.minecraftforge.items.IItemHandler createUnSidedHandler() {
+      return new net.minecraftforge.items.VanillaHopperItemHandler(this);
+   }
+
+   public long getLastUpdateTime() {
+      return this.field_190578_g;
    }
 }
