--- a/net/minecraft/world/border/WorldBorder.java
+++ b/net/minecraft/world/border/WorldBorder.java
@@ -28,6 +_,19 @@
    private WorldBorder.IBorderInfo field_212674_i = new WorldBorder.StationaryBorderInfo(6.0E7D);
    public static final WorldBorder.Serializer field_235925_b_ = new WorldBorder.Serializer(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, 6.0E7D, 0L, 0.0D);
 
+
+   // Paper start
+   private final BlockPos.Mutable mutPos = new BlockPos.Mutable();
+   public boolean isBlockInBounds(int chunkX, int chunkZ) {
+      this.mutPos.func_181079_c(chunkX, 64, chunkZ);
+      return this.func_177746_a(this.mutPos);
+   }
+   public boolean isChunkInBounds(int chunkX, int chunkZ) {
+      this.mutPos.func_181079_c(((chunkX << 4) + 15), 64, (chunkZ << 4) + 15);
+      return this.func_177746_a(this.mutPos);
+   }
+   // Paper end
+
    public boolean func_177746_a(BlockPos p_177746_1_) {
       return (double)(p_177746_1_.func_177958_n() + 1) > this.func_177726_b() && (double)p_177746_1_.func_177958_n() < this.func_177728_d() && (double)(p_177746_1_.func_177952_p() + 1) > this.func_177736_c() && (double)p_177746_1_.func_177952_p() < this.func_177733_e();
    }
@@ -134,6 +_,10 @@
 
    public void func_177737_a(IBorderListener p_177737_1_) {
       this.field_177758_a.add(p_177737_1_);
+   }
+
+   public void removeListener(IBorderListener listener) {
+      this.field_177758_a.remove(listener);
    }
 
    public void func_177725_a(int p_177725_1_) {
