--- a/net/minecraft/entity/monster/DrownedEntity.java
+++ b/net/minecraft/entity/monster/DrownedEntity.java
@@ -77,7 +_,7 @@
       this.field_70714_bg.func_75776_a(7, new RandomWalkingGoal(this, 1.0D));
       this.field_70715_bh.func_75776_a(1, (new HurtByTargetGoal(this, DrownedEntity.class)).func_220794_a(ZombifiedPiglinEntity.class));
       this.field_70715_bh.func_75776_a(2, new NearestAttackableTargetGoal<>(this, PlayerEntity.class, 10, true, false, this::func_204714_e));
-      this.field_70715_bh.func_75776_a(3, new NearestAttackableTargetGoal<>(this, AbstractVillagerEntity.class, false));
+      if ( field_70170_p.spigotConfig.zombieAggressiveTowardsVillager ) this.field_70715_bh.func_75776_a(3, new NearestAttackableTargetGoal<>(this, AbstractVillagerEntity.class, false));
       this.field_70715_bh.func_75776_a(3, new NearestAttackableTargetGoal<>(this, IronGolemEntity.class, true));
       this.field_70715_bh.func_75776_a(5, new NearestAttackableTargetGoal<>(this, TurtleEntity.class, 10, true, false, TurtleEntity.field_203029_bx));
    }
