--- a/net/minecraft/world/biome/provider/EndBiomeProvider.java
+++ b/net/minecraft/world/biome/provider/EndBiomeProvider.java
@@ -29,6 +_,16 @@
    private final Biome field_242643_k;
    private final Biome field_242644_l;
    private final Biome field_242645_m;
+   // Paper start
+   private static final class NoiseCache {
+      public long[] keys = new long[8192];
+      public float[] values = new float[8192];
+      public NoiseCache() {
+         java.util.Arrays.fill(keys, Long.MIN_VALUE);
+      }
+   }
+   private static final ThreadLocal<java.util.Map<SimplexNoiseGenerator, NoiseCache>> noiseCache = ThreadLocal.withInitial(java.util.WeakHashMap::new);
+   // Paper end
 
    public EndBiomeProvider(Registry<Biome> p_i241959_1_, long p_i241959_2_) {
       this(p_i241959_1_, p_i241959_2_, p_i241959_1_.func_243576_d(Biomes.field_76779_k), p_i241959_1_.func_243576_d(Biomes.field_201938_R), p_i241959_1_.func_243576_d(Biomes.field_201937_Q), p_i241959_1_.func_243576_d(Biomes.field_201936_P), p_i241959_1_.func_243576_d(Biomes.field_201939_S));
@@ -83,13 +_,18 @@
       int j = p_235317_2_ / 2;
       int k = p_235317_1_ % 2;
       int l = p_235317_2_ % 2;
-      float f = 100.0F - MathHelper.func_76129_c((float)(p_235317_1_ * p_235317_1_ + p_235317_2_ * p_235317_2_)) * 8.0F;
+      // Paper start - cast ints to long to avoid integer overflow
+      float f = 100.0F - MathHelper.func_76129_c((long) p_235317_1_ * (long) p_235317_1_ + (long) p_235317_2_ * (long) p_235317_2_) * 8.0F;
+      //float f = 100.0F - MathHelper.sqrt((float)(p_235317_1_ * p_235317_1_ + p_235317_2_ * p_235317_2_)) * 8.0F;
+      // Paper end
       f = MathHelper.func_76131_a(f, -100.0F, 80.0F);
 
+      NoiseCache cache = noiseCache.get().computeIfAbsent(p_235317_0_, m -> new NoiseCache()); // Paper
       for(int i1 = -12; i1 <= 12; ++i1) {
          for(int j1 = -12; j1 <= 12; ++j1) {
             long k1 = (long)(i + i1);
             long l1 = (long)(j + j1);
+            //TODO: use cache
             if (k1 * k1 + l1 * l1 > 4096L && p_235317_0_.func_151605_a((double)k1, (double)l1) < (double)-0.9F) {
                float f1 = (MathHelper.func_76135_e((float)k1) * 3439.0F + MathHelper.func_76135_e((float)l1) * 147.0F) % 13.0F + 9.0F;
                float f2 = (float)(k - i1 * 2);
