--- a/net/minecraft/entity/ai/brain/task/SleepAtHomeTask.java
+++ b/net/minecraft/entity/ai/brain/task/SleepAtHomeTask.java
@@ -39,7 +_,8 @@
                }
             }
 
-            BlockState blockstate = p_212832_1_.func_180495_p(globalpos.func_218180_b());
+            BlockState blockstate = p_212832_1_.getTypeIfLoaded(globalpos.func_218180_b()); // Paper
+            if (blockstate == null) { return false; } // Paper
             return globalpos.func_218180_b().func_218137_a(p_212832_2_.func_213303_ch(), 2.0D) && blockstate.func_177230_c().func_203417_a(BlockTags.field_219747_F) && !blockstate.func_177229_b(BedBlock.field_176471_b);
          }
       }
