--- a/net/minecraft/world/lighting/NibbleArrayRepeater.java
+++ b/net/minecraft/world/lighting/NibbleArrayRepeater.java
@@ -9,7 +_,7 @@
 
    public NibbleArrayRepeater(NibbleArray p_i51297_1_, int p_i51297_2_) {
       super(128);
-      System.arraycopy(p_i51297_1_.func_177481_a(), p_i51297_2_ * 128, this.field_76585_a, 0, 128);
+      System.arraycopy(p_i51297_1_.getIfSet(), p_i51297_2_ * 128, this.field_76585_a, 0, 128); // Paper
    }
 
    protected int func_177483_b(int p_177483_1_, int p_177483_2_, int p_177483_3_) {
@@ -17,7 +_,7 @@
    }
 
    public byte[] func_177481_a() {
-      byte[] abyte = new byte[2048];
+      byte[] abyte = BYTE_2048.acquire(); // Paper
 
       for(int i = 0; i < 16; ++i) {
          System.arraycopy(this.field_76585_a, 0, abyte, i * 128, 128);
