--- a/net/minecraft/world/lighting/SkyLightStorage.java
+++ b/net/minecraft/world/lighting/SkyLightStorage.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.lighting;
 
+import com.destroystokyo.paper.util.map.QueuedChangesMapLong2Int;
+import com.destroystokyo.paper.util.map.QueuedChangesMapLong2Object;
 import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
@@ -21,45 +_,50 @@
    private volatile boolean field_215553_p;
 
    protected SkyLightStorage(IChunkLightProvider p_i51288_1_) {
-      super(LightType.SKY, p_i51288_1_, new SkyLightStorage.StorageMap(new Long2ObjectOpenHashMap<>(), new Long2IntOpenHashMap(), Integer.MAX_VALUE));
+      super(LightType.SKY, p_i51288_1_, new SkyLightStorage.StorageMap(new QueuedChangesMapLong2Object<>(), new QueuedChangesMapLong2Int(), Integer.MAX_VALUE, false));
    }
 
    protected int func_215525_d(long p_215525_1_) {
       long i = SectionPos.func_218162_e(p_215525_1_);
       int j = SectionPos.func_218144_c(i);
-      SkyLightStorage.StorageMap skylightstorage$storagemap = this.field_215538_e;
-      int k = skylightstorage$storagemap.field_215653_c.get(SectionPos.func_218169_f(i));
-      if (k != skylightstorage$storagemap.field_215652_b && j < k) {
-         NibbleArray nibblearray = this.func_215531_a(skylightstorage$storagemap, i);
-         if (nibblearray == null) {
-            for(p_215525_1_ = BlockPos.func_218288_f(p_215525_1_); nibblearray == null; nibblearray = this.func_215531_a(skylightstorage$storagemap, i)) {
-               i = SectionPos.func_218172_a(i, Direction.UP);
-               ++j;
-               if (j >= k) {
-                  return 15;
+
+      synchronized (this.visibleUpdateLock) {
+
+         SkyLightStorage.StorageMap skylightstorage$storagemap = this.field_215538_e;
+         int k = skylightstorage$storagemap.field_215653_c.getVisibleAsync(SectionPos.func_218169_f(i));
+         if (k != skylightstorage$storagemap.field_215652_b && j < k) {
+            NibbleArray nibblearray = this.func_215531_a(skylightstorage$storagemap, i);
+            if (nibblearray == null) {
+               for(p_215525_1_ = BlockPos.func_218288_f(p_215525_1_); nibblearray == null; nibblearray = this.func_215531_a(skylightstorage$storagemap, i)) {
+                  i = SectionPos.func_218172_a(i, Direction.UP);
+                  ++j;
+                  if (j >= k) {
+                     return 15;
+                  }
+
+                  p_215525_1_ = BlockPos.func_218291_a(p_215525_1_, 0, 16, 0);
                }
-
-               p_215525_1_ = BlockPos.func_218291_a(p_215525_1_, 0, 16, 0);
             }
+
+            return nibblearray.func_76582_a(SectionPos.func_218171_b(BlockPos.func_218290_b(p_215525_1_)), SectionPos.func_218171_b(BlockPos.func_218274_c(p_215525_1_)), SectionPos.func_218171_b(BlockPos.func_218282_d(p_215525_1_)));
+         } else {
+            return 15;
          }
-
-         return nibblearray.func_76582_a(SectionPos.func_218171_b(BlockPos.func_218290_b(p_215525_1_)), SectionPos.func_218171_b(BlockPos.func_218274_c(p_215525_1_)), SectionPos.func_218171_b(BlockPos.func_218282_d(p_215525_1_)));
-      } else {
-         return 15;
       }
+
    }
 
    protected void func_215524_j(long p_215524_1_) {
       int i = SectionPos.func_218144_c(p_215524_1_);
       if ((this.field_215539_f).field_215652_b > i) {
          (this.field_215539_f).field_215652_b = i;
-         (this.field_215539_f).field_215653_c.defaultReturnValue((this.field_215539_f).field_215652_b);
+         (this.field_215539_f).field_215653_c.queueDefaultReturnValue((this.field_215539_f).field_215652_b);
       }
 
       long j = SectionPos.func_218169_f(p_215524_1_);
-      int k = (this.field_215539_f).field_215653_c.get(j);
+      int k = (this.field_215539_f).field_215653_c.getUpdating(j);
       if (k < i + 1) {
-         (this.field_215539_f).field_215653_c.put(j, i + 1);
+         (this.field_215539_f).field_215653_c.queueUpdate(j, i + 1);
          if (this.field_215558_o.contains(j)) {
             this.func_223404_q(p_215524_1_);
             if (k > (this.field_215539_f).field_215652_b) {
@@ -95,19 +_,19 @@
       }
 
       int j = SectionPos.func_218144_c(p_215523_1_);
-      if ((this.field_215539_f).field_215653_c.get(i) == j + 1) {
+      if ((this.field_215539_f).field_215653_c.getUpdating(i) == j + 1) {
          long k;
          for(k = p_215523_1_; !this.func_215518_g(k) && this.func_215550_a(j); k = SectionPos.func_218172_a(k, Direction.DOWN)) {
             --j;
          }
 
          if (this.func_215518_g(k)) {
-            (this.field_215539_f).field_215653_c.put(i, j + 1);
+            (this.field_215539_f).field_215653_c.queueUpdate(i, j + 1);
             if (flag) {
                this.func_223404_q(k);
             }
          } else {
-            (this.field_215539_f).field_215653_c.remove(i);
+            (this.field_215539_f).field_215653_c.queueRemove(i);
          }
       }
 
@@ -120,7 +_,7 @@
    protected void func_215526_b(long p_215526_1_, boolean p_215526_3_) {
       this.func_215532_c();
       if (p_215526_3_ && this.field_215558_o.add(p_215526_1_)) {
-         int i = (this.field_215539_f).field_215653_c.get(p_215526_1_);
+         int i = (this.field_215539_f).field_215653_c.getUpdating(p_215526_1_);
          if (i != (this.field_215539_f).field_215652_b) {
             long j = SectionPos.func_218166_b(SectionPos.func_218173_b(p_215526_1_), i - 1, SectionPos.func_218153_d(p_215526_1_));
             this.func_223404_q(j);
@@ -142,16 +_,16 @@
          return nibblearray;
       } else {
          long i = SectionPos.func_218172_a(p_215530_1_, Direction.UP);
-         int j = (this.field_215539_f).field_215653_c.get(SectionPos.func_218169_f(p_215530_1_));
+         int j = (this.field_215539_f).field_215653_c.getUpdating(SectionPos.func_218169_f(p_215530_1_));
          if (j != (this.field_215539_f).field_215652_b && SectionPos.func_218144_c(i) < j) {
             NibbleArray nibblearray1;
-            while((nibblearray1 = this.func_215520_a(i, true)) == null) {
+            while((nibblearray1 = this.field_215539_f.getUpdatingOptimized(i)) == null) {
                i = SectionPos.func_218172_a(i, Direction.UP);
             }
 
-            return new NibbleArray((new NibbleArrayRepeater(nibblearray1, 0)).func_177481_a());
+            return new NibbleArray().markPoolSafe(new NibbleArrayRepeater(nibblearray1, 0).func_177481_a()); // Paper - mark pool use as safe (no auto cleaner)
          } else {
-            return new NibbleArray();
+            return new NibbleArray().markPoolSafe(); // Paper - mark pool use as safe (no auto cleaner)
          }
       }
    }
@@ -169,7 +_,7 @@
                         this.field_215539_f.func_215641_a(i);
                      }
 
-                     Arrays.fill(this.func_215520_a(i, true).func_177481_a(), (byte)-1);
+                     Arrays.fill(this.field_215539_f.getUpdatingOptimized(i).asBytesPoolSafe(), (byte)-1); // Paper - use optimized
                      int i3 = SectionPos.func_218142_c(SectionPos.func_218173_b(i));
                      int k3 = SectionPos.func_218142_c(SectionPos.func_218144_c(i));
                      int i4 = SectionPos.func_218142_c(SectionPos.func_218153_d(i));
@@ -257,7 +_,7 @@
          if (!this.field_215558_o.contains(k)) {
             return false;
          } else {
-            int l = (this.field_215539_f).field_215653_c.get(k);
+            int l = (this.field_215539_f).field_215653_c.getUpdating(k);
             return SectionPos.func_218142_c(l) == i + 16;
          }
       }
@@ -265,7 +_,7 @@
 
    protected boolean func_215549_m(long p_215549_1_) {
       long i = SectionPos.func_218169_f(p_215549_1_);
-      int j = (this.field_215539_f).field_215653_c.get(i);
+      int j = (this.field_215539_f).field_215653_c.getUpdating(i);
       return j == (this.field_215539_f).field_215652_b || SectionPos.func_218144_c(p_215549_1_) >= j;
    }
 
@@ -276,17 +_,17 @@
 
    public static final class StorageMap extends LightDataMap<SkyLightStorage.StorageMap> {
       private int field_215652_b;
-      private final Long2IntOpenHashMap field_215653_c;
+      private final QueuedChangesMapLong2Int field_215653_c;
 
-      public StorageMap(Long2ObjectOpenHashMap<NibbleArray> p_i50496_1_, Long2IntOpenHashMap p_i50496_2_, int p_i50496_3_) {
-         super(p_i50496_1_);
+      public StorageMap(QueuedChangesMapLong2Object<NibbleArray> p_i50496_1_, QueuedChangesMapLong2Int p_i50496_2_, int p_i50496_3_, boolean isVisible) {
+         super(p_i50496_1_, isVisible);
          this.field_215653_c = p_i50496_2_;
-         p_i50496_2_.defaultReturnValue(p_i50496_3_);
+         p_i50496_2_.queueDefaultReturnValue(p_i50496_3_);
          this.field_215652_b = p_i50496_3_;
       }
 
       public SkyLightStorage.StorageMap func_212858_b_() {
-         return new SkyLightStorage.StorageMap(this.field_215645_a.clone(), this.field_215653_c.clone(), this.field_215652_b);
+         return new SkyLightStorage.StorageMap(this.field_215645_a, this.field_215653_c, this.field_215652_b, true);
       }
    }
 }
