--- a/net/minecraft/entity/AgeableEntity.java
+++ b/net/minecraft/entity/AgeableEntity.java
@@ -17,6 +_,31 @@
    protected int field_175502_b;
    protected int field_175503_c;
 
+   // Spigot start
+   @Override
+   public void inactiveTick()
+   {
+      super.inactiveTick();
+      if ( this.field_70170_p.field_72995_K )
+      { // CraftBukkit
+         this.func_213323_x_();
+      } else
+      {
+         int i = this.func_70874_b();
+
+         if ( i < 0 )
+         {
+            ++i;
+            this.func_70873_a( i );
+         } else if ( i > 0 )
+         {
+            --i;
+            this.func_70873_a( i );
+         }
+      }
+   }
+   // Spigot end
+
    protected AgeableEntity(EntityType<? extends AgeableEntity> p_i48581_1_, World p_i48581_2_) {
       super(p_i48581_1_, p_i48581_2_);
    }
