--- a/net/minecraft/network/play/client/CChatMessagePacket.java
+++ b/net/minecraft/network/play/client/CChatMessagePacket.java
@@ -4,6 +_,7 @@
 import net.minecraft.network.IPacket;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.IServerPlayNetHandler;
+import org.goldenforge.threads.GoldenForgeExecutors;
 
 public class CChatMessagePacket implements IPacket<IServerPlayNetHandler> {
    private String field_149440_a;
@@ -28,6 +_,11 @@
    }
 
    public void func_148833_a(IServerPlayNetHandler p_148833_1_) {
+      if ( !field_149440_a.startsWith("/") )
+      {
+         GoldenForgeExecutors.CHAT.submit(() -> p_148833_1_.func_147354_a(CChatMessagePacket.this));
+         return;
+      }
       p_148833_1_.func_147354_a(this);
    }
 
