--- a/net/minecraft/util/math/BlockPos.java
+++ b/net/minecraft/util/math/BlockPos.java
@@ -60,28 +_,29 @@
       this(p_i46034_1_.func_177958_n(), p_i46034_1_.func_177956_o(), p_i46034_1_.func_177952_p());
    }
 
+   public static long getAdjacent(int baseX, int baseY, int baseZ, Direction enumdirection) { return func_218276_a(baseX + enumdirection.getAdjacentX(), baseY + enumdirection.getAdjacentY(), baseZ + enumdirection.getAdjacentZ()); } // Paper
    public static long func_218289_a(long p_218289_0_, Direction p_218289_2_) {
       return func_218291_a(p_218289_0_, p_218289_2_.func_82601_c(), p_218289_2_.func_96559_d(), p_218289_2_.func_82599_e());
    }
 
    public static long func_218291_a(long p_218291_0_, int p_218291_2_, int p_218291_3_, int p_218291_4_) {
-      return func_218276_a(func_218290_b(p_218291_0_) + p_218291_2_, func_218274_c(p_218291_0_) + p_218291_3_, func_218282_d(p_218291_0_) + p_218291_4_);
+      return func_218276_a((int) (p_218291_0_ >> 38) + p_218291_2_, (int) ((p_218291_0_ << 52) >> 52) + p_218291_3_, (int) ((p_218291_0_ << 26) >> 38) + p_218291_4_); // Paper - simplify/inline
    }
 
    public static int func_218290_b(long p_218290_0_) {
-      return (int)(p_218290_0_ << 64 - field_218293_k - field_177990_b >> 64 - field_177990_b);
+      return (int) (p_218290_0_ >> 38); // Paper - simplify/inline
    }
 
    public static int func_218274_c(long p_218274_0_) {
-      return (int)(p_218274_0_ << 64 - field_177989_d >> 64 - field_177989_d);
+      return (int) ((p_218274_0_ << 52) >> 52); // Paper - simplify/inline
    }
 
    public static int func_218282_d(long p_218282_0_) {
-      return (int)(p_218282_0_ << 64 - field_218292_j - field_177991_c >> 64 - field_177991_c);
+      return (int) ((p_218282_0_ << 26) >> 38);  // Paper - simplify/inline
    }
 
    public static BlockPos func_218283_e(long p_218283_0_) {
-      return new BlockPos(func_218290_b(p_218283_0_), func_218274_c(p_218283_0_), func_218282_d(p_218283_0_));
+      return new BlockPos((int) (p_218283_0_ >> 38), (int) ((p_218283_0_ << 52) >> 52), (int) ((p_218283_0_ << 26) >> 38)); // Paper - simplify/inline
    }
 
    public long func_218275_a() {
@@ -89,10 +_,12 @@
    }
 
    public static long func_218276_a(int p_218276_0_, int p_218276_1_, int p_218276_2_) {
-      long i = 0L;
-      i = i | ((long)p_218276_0_ & field_177994_h) << field_218293_k;
-      i = i | ((long)p_218276_1_ & field_177995_i) << 0;
-      return i | ((long)p_218276_2_ & field_177993_j) << field_218292_j;
+      return (((long) p_218276_0_ & (long) 67108863) << 38) | (((long) p_218276_1_ & (long) 4095)) | (((long) p_218276_2_ & (long) 67108863) << 12); // Paper - inline constants and simplify
+
+//      long i = 0L;
+//      i = i | ((long)p_218276_0_ & PACKED_X_MASK) << X_OFFSET;
+//      i = i | ((long)p_218276_1_ & PACKED_Y_MASK) << 0;
+//      return i | ((long)p_218276_2_ & PACKED_Z_MASK) << Z_OFFSET;
    }
 
    public static long func_218288_f(long p_218288_0_) {
@@ -116,55 +_,73 @@
    }
 
    public BlockPos func_177984_a() {
-      return this.func_177972_a(Direction.UP);
+      return new BlockPos(this.func_177958_n(), this.func_177956_o() + 1, this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177981_b(int p_177981_1_) {
-      return this.func_177967_a(Direction.UP, p_177981_1_);
+      return p_177981_1_ == 0 ? this : new BlockPos(this.func_177958_n(), this.func_177956_o() + p_177981_1_, this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177977_b() {
-      return this.func_177972_a(Direction.DOWN);
+      return new BlockPos(this.func_177958_n(), this.func_177956_o() - 1, this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177979_c(int p_177979_1_) {
-      return this.func_177967_a(Direction.DOWN, p_177979_1_);
+      return p_177979_1_ == 0 ? this : new BlockPos(this.func_177958_n(), this.func_177956_o() - p_177979_1_, this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177978_c() {
-      return this.func_177972_a(Direction.NORTH);
+      return new BlockPos(this.func_177958_n(), this.func_177956_o(), this.func_177952_p() - 1); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177964_d(int p_177964_1_) {
-      return this.func_177967_a(Direction.NORTH, p_177964_1_);
+      return p_177964_1_ == 0 ? this : new BlockPos(this.func_177958_n(), this.func_177956_o(), this.func_177952_p() - p_177964_1_); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177968_d() {
-      return this.func_177972_a(Direction.SOUTH);
+      return new BlockPos(this.func_177958_n(), this.func_177956_o(), this.func_177952_p() + 1); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177970_e(int p_177970_1_) {
-      return this.func_177967_a(Direction.SOUTH, p_177970_1_);
+      return p_177970_1_ == 0 ? this : new BlockPos(this.func_177958_n(), this.func_177956_o(), this.func_177952_p() + p_177970_1_); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177976_e() {
-      return this.func_177972_a(Direction.WEST);
+      return new BlockPos(this.func_177958_n() - 1, this.func_177956_o(), this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177985_f(int p_177985_1_) {
-      return this.func_177967_a(Direction.WEST, p_177985_1_);
+      return p_177985_1_ == 0 ? this : new BlockPos(this.func_177958_n() - p_177985_1_, this.func_177956_o(), this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177974_f() {
-      return this.func_177972_a(Direction.EAST);
+      return new BlockPos(this.func_177958_n() + 1, this.func_177956_o(), this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177965_g(int p_177965_1_) {
-      return this.func_177967_a(Direction.EAST, p_177965_1_);
+      return p_177965_1_ == 0 ? this : new BlockPos(this.func_177958_n() + p_177965_1_, this.func_177956_o(), this.func_177952_p()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos func_177972_a(Direction p_177972_1_) {
-      return new BlockPos(this.func_177958_n() + p_177972_1_.func_82601_c(), this.func_177956_o() + p_177972_1_.func_96559_d(), this.func_177952_p() + p_177972_1_.func_82599_e());
+      //return new BlockPos(this.getX() + p_177972_1_.getStepX(), this.getY() + p_177972_1_.getStepY(), this.getZ() + p_177972_1_.getStepZ());
+      // Paper Start - Optimize BlockPosition
+      switch(p_177972_1_) {
+         case UP:
+            return new BlockPos(this.func_177958_n(), this.func_177956_o() + 1, this.func_177952_p());
+         case DOWN:
+            return new BlockPos(this.func_177958_n(), this.func_177956_o() - 1, this.func_177952_p());
+         case NORTH:
+            return new BlockPos(this.func_177958_n(), this.func_177956_o(), this.func_177952_p() - 1);
+         case SOUTH:
+            return new BlockPos(this.func_177958_n(), this.func_177956_o(), this.func_177952_p() + 1);
+         case WEST:
+            return new BlockPos(this.func_177958_n() - 1, this.func_177956_o(), this.func_177952_p());
+         case EAST:
+            return new BlockPos(this.func_177958_n() + 1, this.func_177956_o(), this.func_177952_p());
+         default:
+            return new BlockPos(this.func_177958_n() + p_177972_1_.func_82601_c(), this.func_177956_o() + p_177972_1_.func_96559_d(), this.func_177952_p() + p_177972_1_.func_82599_e());
+      }
+      // Paper End
    }
 
    public BlockPos func_177967_a(Direction p_177967_1_, int p_177967_2_) {
@@ -468,17 +_,17 @@
          }
       }
 
-      public void func_223471_o(int p_223471_1_) {
-         super.func_223471_o(p_223471_1_);
-      }
-
-      public void func_185336_p(int p_185336_1_) {
-         super.func_185336_p(p_185336_1_);
-      }
-
-      public void func_223472_q(int p_223472_1_) {
-         super.func_223472_q(p_223472_1_);
-      }
+//      public void setX(int p_223471_1_) {
+//         super.setX(p_223471_1_);
+//      }
+//
+//      public void setY(int p_185336_1_) {
+//         super.setY(p_185336_1_);
+//      }
+//
+//      public void setZ(int p_223472_1_) {
+//         super.setZ(p_223472_1_);
+//      }
 
       public BlockPos func_185334_h() {
          return new BlockPos(this);
