--- a/net/minecraft/util/math/SectionPos.java
+++ b/net/minecraft/util/math/SectionPos.java
@@ -29,6 +_,10 @@
       return new SectionPos(func_218159_a(MathHelper.func_76128_c(p_218157_0_.func_226277_ct_())), func_218159_a(MathHelper.func_76128_c(p_218157_0_.func_226278_cu_())), func_218159_a(MathHelper.func_76128_c(p_218157_0_.func_226281_cx_())));
    }
 
+   public static SectionPos a(long i) {
+      return new SectionPos((int) (i >> 42), (int) (i << 44 >> 44), (int) (i << 22 >> 42)); // Paper
+   }
+
    public static SectionPos func_218170_a(long p_218170_0_) {
       return new SectionPos(func_218173_b(p_218170_0_), func_218144_c(p_218170_0_), func_218153_d(p_218170_0_));
    }
@@ -37,6 +_,12 @@
       return func_218174_a(p_218172_0_, p_218172_2_.func_82601_c(), p_218172_2_.func_96559_d(), p_218172_2_.func_82599_e());
    }
 
+   // Paper end
+   public static long a(long i, int j, int k, int l) {
+      return (((long) ((int) (i >> 42) + j) & 4194303L) << 42) | (((long) ((int) (i << 44 >> 44) + k) & 1048575L)) | (((long) ((int) (i << 22 >> 42) + l) & 4194303L) << 20); // Simplify to reduce instruction count
+   }
+
+
    public static long func_218174_a(long p_218174_0_, int p_218174_2_, int p_218174_3_, int p_218174_4_) {
       return func_218166_b(func_218173_b(p_218174_0_) + p_218174_2_, func_218144_c(p_218174_0_) + p_218174_3_, func_218153_d(p_218174_0_) + p_218174_4_);
    }
@@ -50,10 +_,8 @@
    }
 
    public static short func_218150_b(BlockPos p_218150_0_) {
-      int i = func_218171_b(p_218150_0_.func_177958_n());
-      int j = func_218171_b(p_218150_0_.func_177956_o());
-      int k = func_218171_b(p_218150_0_.func_177952_p());
-      return (short)(i << 8 | k << 4 | j << 0);
+      return (short) ((p_218150_0_.func_177958_n() & 15) << 8 | (p_218150_0_.func_177952_p() & 15) << 4 | p_218150_0_.func_177956_o() & 15); // Paper - simplify/inline
+
    }
 
    public static int func_243641_a(short p_243641_0_) {
@@ -100,6 +_,14 @@
       return (int)(p_218153_0_ << 22 >> 42);
    }
 
+   public static long getAdjacentFromBlockPos(int x, int y, int z, Direction enumdirection) {
+      return (((long) ((x >> 4) + enumdirection.getAdjacentX()) & 4194303L) << 42) | (((long) ((y >> 4) + enumdirection.getAdjacentY()) & 1048575L)) | (((long) ((z >> 4) + enumdirection.getAdjacentZ()) & 4194303L) << 20);
+   }
+   public static long getAdjacentFromSectionPos(int x, int y, int z, Direction enumdirection) {
+      return (((long) (x + enumdirection.getAdjacentX()) & 4194303L) << 42) | (((long) ((y) + enumdirection.getAdjacentY()) & 1048575L)) | (((long) (z + enumdirection.getAdjacentZ()) & 4194303L) << 20);
+   }
+   // Paper end
+
    public int func_218149_a() {
       return this.func_177958_n();
    }
@@ -158,15 +_,18 @@
    }
 
    public static long func_218166_b(int p_218166_0_, int p_218166_1_, int p_218166_2_) {
-      long i = 0L;
-      i = i | ((long)p_218166_0_ & 4194303L) << 42;
-      i = i | ((long)p_218166_1_ & 1048575L) << 0;
-      return i | ((long)p_218166_2_ & 4194303L) << 20;
+      return (((long) p_218166_0_ & 4194303L) << 42) | (((long) p_218166_1_ & 1048575L)) | (((long) p_218166_2_ & 4194303L) << 20); // Paper - Simplify to reduce instruction count
    }
 
    public long func_218146_v() {
-      return func_218166_b(this.func_218149_a(), this.func_218163_b(), this.func_218148_c());
-   }
+      return (((long) func_177958_n() & 4194303L) << 42) | (((long) func_177956_o() & 1048575L)) | (((long) func_177952_p() & 4194303L) << 20); // Paper - Simplify to reduce instruction count
+   }
+
+   // Paper start
+   public static long blockPosAsSectionLong(int i, int j, int k) {
+      return (((long) (i >> 4) & 4194303L) << 42) | (((long) (j >> 4) & 1048575L)) | (((long) (k >> 4) & 4194303L) << 20);
+   }
+   // Paper end
 
    public Stream<BlockPos> func_218145_w() {
       return BlockPos.func_218287_a(this.func_218161_d(), this.func_218151_e(), this.func_218164_f(), this.func_218152_g(), this.func_218165_h(), this.func_218143_r());
