--- a/net/minecraft/entity/passive/SquidEntity.java
+++ b/net/minecraft/entity/passive/SquidEntity.java
@@ -46,7 +_,7 @@
 
    public SquidEntity(EntityType<? extends SquidEntity> p_i50243_1_, World p_i50243_2_) {
       super(p_i50243_1_, p_i50243_2_);
-      this.field_70146_Z.setSeed((long)this.func_145782_y());
+      //this.random.setSeed((long)this.getId());
       this.field_70864_bA = 1.0F / (this.field_70146_Z.nextFloat() + 1.0F) * 0.2F;
    }
 
@@ -178,7 +_,9 @@
    }
 
    public static boolean func_223365_b(EntityType<SquidEntity> p_223365_0_, IWorld p_223365_1_, SpawnReason p_223365_2_, BlockPos p_223365_3_, Random p_223365_4_) {
-      return p_223365_3_.func_177956_o() > 45 && p_223365_3_.func_177956_o() < p_223365_1_.func_181545_F();
+      final double maxHeight = ((ServerWorld)p_223365_1_).paperConfig.squidMaxSpawnHeight > 0 ? ((ServerWorld)p_223365_1_).paperConfig.squidMaxSpawnHeight : p_223365_1_.func_181545_F(); // Paper
+      return p_223365_3_.func_177956_o() > ((ServerWorld)p_223365_1_).spigotConfig.squidSpawnRangeMin && p_223365_3_.func_177956_o() < maxHeight; // Spigot // Paper
+      //return p_223365_3_.getY() > ((ServerWorld)p_223365_1_).spigotConfig.squidSpawnRangeMin && p_223365_3_.getY() < p_223365_1_.getSeaLevel();
    }
 
    @OnlyIn(Dist.CLIENT)
