--- a/net/minecraft/entity/passive/StriderEntity.java
+++ b/net/minecraft/entity/passive/StriderEntity.java
@@ -395,12 +_,12 @@
       } else {
          Object object;
          if (this.field_70146_Z.nextInt(30) == 0) {
-            MobEntity mobentity = EntityType.field_233592_ba_.func_200721_a(p_213386_1_.func_201672_e());
+            MobEntity mobentity = EntityType.field_233592_ba_.func_200721_a(p_213386_1_.getLevel());
             object = this.func_242331_a(p_213386_1_, p_213386_2_, mobentity, new ZombieEntity.GroupData(ZombieEntity.func_241399_a_(this.field_70146_Z), false));
             mobentity.func_184201_a(EquipmentSlotType.MAINHAND, new ItemStack(Items.field_234774_pk_));
             this.func_230266_a_((SoundCategory)null);
          } else if (this.field_70146_Z.nextInt(10) == 0) {
-            AgeableEntity ageableentity = EntityType.field_233589_aE_.func_200721_a(p_213386_1_.func_201672_e());
+            AgeableEntity ageableentity = EntityType.field_233589_aE_.func_200721_a(p_213386_1_.getLevel());
             ageableentity.func_70873_a(-24000);
             object = this.func_242331_a(p_213386_1_, p_213386_2_, ageableentity, (ILivingEntityData)null);
          } else {
