--- a/net/minecraft/world/spawner/WanderingTraderSpawner.java
+++ b/net/minecraft/world/spawner/WanderingTraderSpawner.java
@@ -96,7 +_,7 @@
                return false;
             }
 
-            WanderingTraderEntity wanderingtraderentity = EntityType.field_220351_aK.func_220342_a(p_234562_1_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, blockpos2, SpawnReason.EVENT, false, false);
+            WanderingTraderEntity wanderingtraderentity = EntityType.field_220351_aK.spawn(p_234562_1_, (CompoundNBT)null, (ITextComponent)null, (PlayerEntity)null, blockpos2, SpawnReason.EVENT, false, false, SpawnReason.NATURAL);
             if (wanderingtraderentity != null) {
                for(int j = 0; j < 2; ++j) {
                   this.func_242373_a(p_234562_1_, wanderingtraderentity, 4);
