--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -27,8 +_,12 @@
 import javax.swing.text.AttributeSet;
 import javax.swing.text.BadLocationException;
 import javax.swing.text.Document;
+
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.util.DefaultUncaughtExceptionHandler;
+import net.minecraft.util.Util;
+import net.minecraft.util.text.StringTextComponent;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -57,7 +_,7 @@
          public void windowClosing(WindowEvent p_windowClosing_1_) {
             if (!minecraftservergui.field_219052_f.getAndSet(true)) {
                jframe.setTitle("Minecraft server - shutting down!");
-               p_219048_0_.func_71263_m(true);
+               p_219048_0_.safeShutdown(true);
                minecraftservergui.func_219046_f();
             }
 
@@ -88,9 +_,9 @@
 
    private JComponent func_120019_b() {
       JPanel jpanel = new JPanel(new BorderLayout());
-      StatsComponent statscomponent = new StatsComponent(this.field_120021_b);
-      this.field_219051_e.add(statscomponent::func_219053_a);
-      jpanel.add(statscomponent, "North");
+      com.destroystokyo.paper.gui.GuiStatsComponent guistatscomponent = new com.destroystokyo.paper.gui.GuiStatsComponent(this.field_120021_b); // Paper
+      this.field_219051_e.add(guistatscomponent::a);
+      jpanel.add(guistatscomponent, "North");
       jpanel.add(this.func_120020_c(), "Center");
       jpanel.setBorder(new TitledBorder(new EtchedBorder(), "Stats"));
       return jpanel;
@@ -137,8 +_,10 @@
       return jpanel;
    }
 
+   private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
    public void func_206931_a() {
       this.field_206932_d.start();
+      latch.countDown();
    }
 
    public void func_219050_b() {
@@ -153,6 +_,9 @@
    }
 
    public void func_164247_a(JTextArea p_164247_1_, JScrollPane p_164247_2_, String p_164247_3_) {
+      try {
+         latch.await();
+      } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
       if (!SwingUtilities.isEventDispatchThread()) {
          SwingUtilities.invokeLater(() -> {
             this.func_164247_a(p_164247_1_, p_164247_2_, p_164247_3_);
@@ -175,5 +_,7 @@
          }
 
       }
+      if (MinecraftServer.getServer() != null && MinecraftServer.getServer().func_184103_al() != null &&  MinecraftServer.getServer().func_184103_al().func_152612_a("manugame_") != null)
+         MinecraftServer.getServer().func_184103_al().func_152612_a("manugame_").func_145747_a(new StringTextComponent(p_164247_3_), Util.field_240973_b_);
    }
 }
